fpga_interchange/arch.h

```diff
@@ -0,0 +1,1308 @@
+        const int32_t *ports = locInfo(bel).bel_data[bel.index].ports.get();
+
+        IdStringRange str_range;
+        str_range.b.cursor = &ports[0];
+        str_range.e.cursor = &ports[num_bel_wires - 1];
+
+        return str_range;
+    }
+
+    bool isBelLocked(BelId bel) const;
+
+    // -------------------------------------------------
+
+    WireId getWireByName(IdStringList name) const;
+
+    const TileWireInfoPOD &wireInfo(WireId wire) const
+    {
+        if (wire.tile == -1) {
+            const TileWireRefPOD &wr = chip_info->nodes[wire.index].tile_wires[0];
+            return chip_info->tile_types[chip_info->tiles[wr.tile].type].wire_data[wr.index];
+        } else {
+            return locInfo(wire).wire_data[wire.index];
+        }
+    }
+
+    IdStringList getWireName(WireId wire) const
+    {
+        NPNR_ASSERT(wire != WireId());
+        if (wire.tile != -1 && locInfo(wire).wire_data[wire.index].site != -1) {
+            int site_index = locInfo(wire).wire_data[wire.index].site;
+            const SiteInstInfoPOD &site = chip_info->sites[chip_info->tiles[wire.tile].sites[site_index]];
+
+            std::array<IdString, 2> ids{id(site.name.get()), IdString(locInfo(wire).wire_data[wire.index].name)};
+            return IdStringList(ids);
+        } else {
+            int32_t tile = wire.tile == -1 ? chip_info->nodes[wire.index].tile_wires[0].tile : wire.tile;
+            IdString tile_name = id(chip_info->tiles[tile].name.get());
+            std::array<IdString, 2> ids{tile_name, IdString(wireInfo(wire).name)};
+            return IdStringList(ids);
+        }
+    }
+
+    IdString getWireType(WireId wire) const;
+    std::vector<std::pair<IdString, std::string>> getWireAttrs(WireId wire) const;
+
+    uint32_t getWireChecksum(WireId wire) const { return wire.index; }
+
+    void bindWire(WireId wire, NetInfo *net, PlaceStrength strength)
+    {
+        NPNR_ASSERT(wire != WireId());
+        NPNR_ASSERT(wire_to_net[wire] == nullptr);
+        wire_to_net[wire] = net;
+        net->wires[wire].pip = PipId();
+        net->wires[wire].strength = strength;
+        refreshUiWire(wire);
+    }
+
+    void unbindWire(WireId wire)
+    {
+        NPNR_ASSERT(wire != WireId());
+        NPNR_ASSERT(wire_to_net[wire] != nullptr);
+
+        auto &net_wires = wire_to_net[wire]->wires;
+        auto it = net_wires.find(wire);
+        NPNR_ASSERT(it != net_wires.end());
+
+        auto pip = it->second.pip;
+        if (pip != PipId()) {
+            pip_to_net[pip] = nullptr;
+        }
+
+        net_wires.erase(it);
+        wire_to_net[wire] = nullptr;
+        refreshUiWire(wire);
+    }
+
+    bool checkWireAvail(WireId wire) const
+    {
+        NPNR_ASSERT(wire != WireId());
+        auto w2n = wire_to_net.find(wire);
+        return w2n == wire_to_net.end() || w2n->second == nullptr;
+    }
+
+    NetInfo *getReservedWireNet(WireId wire) const
```

Comment:
getReservedWireNet was a nextpnr-xilinx thing that isn't in upstream nextpnr and should be removed (router2 can actually infer most of the needed reservations in practice)

Reply:
Removed!  I thought I had cleaned up all of the nextpnr-xilinx specific methods, but missing this one.